aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,171 @@
+import inspect
+from pathlib import Path
+from queue import Queue
+from types import SimpleNamespace
+from typing import Optional, Tuple, Type, Union, cast
+
+from aea.context.base import AgentContext
+from aea.identity.base import Identity
+from aea.mail.base import Address
+from aea.multiplexer import AsyncMultiplexer, Multiplexer, OutBox
+from aea.protocols.base import Message
+from aea.skills.base import Skill
+from aea.skills.tasks import TaskManager
+
+
+class BaseSkillTestCase:
+    """A class to test a skill."""
+
+    path_to_skill: Union[Path, str] = Path(".")
+    _skill: Skill
+    _multiplexer: AsyncMultiplexer
+    _outbox: OutBox
+
+    @property
+    def skill(self) -> Skill:
+        """Get the skill."""
+        try:
+            value = self._skill
+        except AttributeError:
+            raise ValueError("Ensure skill is set during setup_class.")
+        return value
+
+    def get_quantity_in_outbox(self) -> int:
+        """Get the quantity of envelopes in the outbox."""
+        return self._multiplexer.out_queue.qsize()
+
+    def get_message_from_outbox(self) -> Optional[Message]:
+        """Get message from outbox."""
+        if self._outbox.empty():
+            return None
+        envelope = self._multiplexer.out_queue.get_nowait()
+        return envelope.message
+
+    @staticmethod
+    def message_has_attributes(
+        actual_message: Message, message_type: Type[Message], **kwargs,
+    ) -> Tuple[bool, str]:
+        """
+        Evaluates whether a message's attributes match the expected attributes provided.
+
+        :param actual_message: the actual message
+        :param message_type: the expected message type
+        :param kwargs: other expected message attributes
+
+        :return: boolean result of the evaluation and accompanied message
+        """
+        if type(actual_message) != message_type:
+            return (
+                False,
+                "The message types do not match. Actual type: {}. Expected type: {}".format(
+                    type(actual_message), message_type
+                ),
+            )
+
+        for attribute_name, expected_value in kwargs.items():
+            attribute = getattr(actual_message, attribute_name)
+            if callable(attribute):
+                if attribute != expected_value:
+                    return (
+                        False,
+                        "The '{}' fields do not match. Actual '{}': {}. Expected '{}': {}".format(
+                            attribute_name,
+                            attribute_name,
+                            attribute,
+                            attribute_name,
+                            expected_value,
+                        ),
+                    )
+
+        return True, "The message has the provided expected attributes."
+
+    def build_incoming_message(
+        self,
+        message_type: Type[Message],
+        dialogue_reference: Optional[Tuple[str, str]] = None,
```

Comment:
An incoming message means that the dialogue is held somewhere outside the skill. So you can simply generate the dialogue reference inside the method rather than have it passed as an argument.

Reply:
Sometimes you want to create an incoming message with a specific incorrect dialogue reference (e.g. testing `_handle_unidentified_dialogue()`). This accommodates this kind of test.

If the dialogue_reference isn't passed, the default ("", "") will be used.

Reply:
The default dialogue reference will not be accepted for addition in the receiving dialogue as it is an invalid reference.

Reply:
okI change it so if no dialogue_reference is provided, a new correct reference is used.